owasp-top10-2021-apps/a1/ecommerce-api/app/handlers/handlers.go

```diff
@@ -16,22 +19,43 @@ func HealthCheck(c echo.Context) error {
 // GetTicket returns the userID ticket.
 func GetTicket(c echo.Context) error {
 	id := c.Param("id")
+	cookie, err := c.Cookie("sessionIDa5")
+	if err != nil {
+        return c.JSON(http.StatusBadRequest, map[string]string{"result": "error", "details": "Cookie Error."})
+    }
 	userDataQuery := map[string]interface{}{"userID": id}
 	userDataResult, err := db.GetUserData(userDataQuery)
 	if err != nil {
 		// could not find this user in MongoDB (or MongoDB err connection)
 		return c.JSON(http.StatusBadRequest, map[string]string{"result": "error", "details": "Error finding this UserID."})
 	}
-
-	format := c.QueryParam("format")
-	if format == "json" {
-		return c.JSON(http.StatusOK, map[string]string{
-			"result":   "success",
-			"username": userDataResult.Username,
-			"ticket":   userDataResult.Ticket,
-		})
+	
+	claims := &types.Claims{}
+	tkn, err := jwt.ParseWithClaims(cookie.Value, claims, func(token *jwt.Token) (interface{}, error) {
+		return []byte(os.Getenv("JWT_KEY")), nil
+	})
+	if err != nil {
+		if err == jwt.ErrSignatureInvalid {
+			return c.JSON(http.StatusUnauthorized, map[string]string{"result": "error", "details": "Error StatusUnauthorized."})
+		}
+		return c.JSON(http.StatusBadRequest, map[string]string{"result": "error", "details": "Error StatusBadRequest."})
 	}
+	if !tkn.Valid {
+		return c.JSON(http.StatusUnauthorized, map[string]string{"result": "error", "details": "Error StatusUnauthorized."})
+	}
+
+	if claims.Username == userDataResult.Username{
```

Comment:
Now that we are getting the username from the JWT, we could remove the username parameter and avoid this check. What do you think?

Reply:
Cool, but we are still reading the user id from the parameters and doing the query afterwards:

```go
id := c.Param("id")
userDataQuery := map[string]interface{}{"userID": id}
```

You could do something such as:

```go
// Parse JWT claims
userID, _ := claims["id"]
userDataQuery := map[string]interface{}{"userID": userID}
```

If you want to try that, take a look at [this PR](https://github.com/globocom/secDevLabs/pull/275).